src/client/contextChannels.ts

```diff
@@ -554,38 +554,81 @@ function deserializeWindowAddedEvent(eventTransport: Transport<ChannelWindowAdde
 }
 
-if (typeof fin !== 'undefined') {
-    getServicePromise().then((channelClient) => {
-        channelClient.register(APIToClientTopic.CHANNEL_RECEIVE_CONTEXT, async (payload: ChannelReceiveContextPayload) => {
-            await invokeListeners(
-                channelContextListeners.filter((listener) => listener.channel.id === payload.channel),
-                payload.context,
-                (e) => console.warn(`Error thrown by channel context handler, swallowing error. Error message: ${e.message}`),
-                () => new Error('All channel context handlers failed')
-            );
-        });
+// Keep track of the channel the client is in so it can be rejoined on disconnect
+let currentChannel: DefaultChannel | SystemChannel | AppChannel | null = null;
 
-        const eventHandler = getEventRouter();
+function deserializeChannelChangedEvent(eventTransport: Transport<ChannelChangedEvent>): ChannelChangedEvent {
+    const type = eventTransport.type;
+    const identity = eventTransport.identity;
+    const channel = eventTransport.channel ? getChannelObject(eventTransport.channel) : null;
+    const previousChannel = eventTransport.previousChannel ? getChannelObject(eventTransport.previousChannel) : null;
+
+    if (fin.Window.me.name === identity.name && fin.Window.me.uuid === identity.uuid) {
+        currentChannel = channel;
+    }
 
+    return {type, identity, channel, previousChannel};
+}
+
+if (typeof fin !== 'undefined') {
+    if (getServiceIdentity().name !== fin.Window.me.name && getServiceIdentity().uuid !== fin.Window.me.uuid) {
+        const eventHandler = getEventRouter();
         eventHandler.registerEmitterProvider('channel', (channelId: ChannelId) => {
             return channelEventEmitters[channelId];
         });
 
+        eventHandler.registerDeserializer('channel-changed', deserializeChannelChangedEvent);
         eventHandler.registerDeserializer('window-added', deserializeWindowAddedEvent);
         eventHandler.registerDeserializer('window-removed', deserializeWindowRemovedEvent);
+
+        initialize().then(() => {
+            registerOnChannelConnect(initialize);
+        });
+    }
+}
+
+function initialize(): Promise<void> {
+    return getServicePromise().then(async (channelClient) => {
+        try {
+            channelClient.register(APIToClientTopic.CHANNEL_RECEIVE_CONTEXT, async (payload: ChannelReceiveContextPayload) => {
+                await invokeListeners(
+                    channelContextListeners.filter((listener) => listener.channel.id === payload.channel),
+                    payload.context,
+                    (e) => console.warn(`Error thrown by channel context handler, swallowing error. Error message: ${e.message}`),
+                    () => new Error('All channel context handlers failed')
+                );
+            });
+        } catch (e) {
+            // Trying to resubscribe to the same channel
+        }
+
+        await rehydrate();
```

Comment:
Should this be on reconnect, rather than first initialization? If you're client is refreshing, I feel like that's on you to handle